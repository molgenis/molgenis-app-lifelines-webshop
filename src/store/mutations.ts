import ApplicationState, { Toast } from '@/types/ApplicationState'
import { Variable, VariableWithVariants } from '@/types/Variable'
import Assessment from '@/types/Assessment'
import Count from '@/types/Count'
import Vue from 'vue'
import GridSelection from '@/types/GridSelection'
import Filter from '@/types/Filter'
import { Section } from '@/types/Section.ts'
import { TreeChild, TreeParentInternal } from '@/types/Tree'
import { Order } from '@/types/Order'

export default {
  setIsSignedIn (state: ApplicationState, isSignedIn: boolean) {
    state.isSignedIn = isSignedIn
  },
  setToast (state: ApplicationState, toast: Toast) {
    state.toast = toast
  },
  clearToast (state: ApplicationState) {
    state.toast = null
  },
  setOrders (state: ApplicationState, orders: Order[]) {
    state.orders = orders
  },
  updateFacetFilter (state: ApplicationState, facetFilter: Filter) {
    state.facetFilter = facetFilter
  },
  updateGenderFilter (state: ApplicationState, selectedGenders: string[]) {
    state.facetFilter.gender = selectedGenders
  },
  updateSubcohortfilter (state: ApplicationState, selectedSubcohorts: string[]) {
    state.facetFilter.subcohort = selectedSubcohorts
  },
  updateSelectedAgeAt (state: ApplicationState, selectedAgeAt: any) {
    state.facetFilter.ageGroupAt1A = selectedAgeAt.ageGroupAt1A
    state.facetFilter.ageGroupAt2A = selectedAgeAt.ageGroupAt2A
    state.facetFilter.ageGroupAt3A = selectedAgeAt.ageGroupAt3A
  },
  updateYearOfBirthRangefilter (state: ApplicationState, yobRange: number[]) {
    state.facetFilter.yearOfBirthRange = yobRange
  },
  removeYearOfBirthRangefilter (state: ApplicationState) {
    state.facetFilter.yearOfBirthRange = []
  },
  removeAgeAtFilter (state: ApplicationState) {
    state.facetFilter.ageGroupAt1A = []
    state.facetFilter.ageGroupAt2A = []
    state.facetFilter.ageGroupAt3A = []
  },
  updateTreeSelection (state: ApplicationState, selection: number) {
    state.treeSelected = selection
  },
  updateTreeOpenSection (state: ApplicationState, treeOpenSection: number) {
    state.treeOpenSection = treeOpenSection
    if (treeOpenSection !== -1) {
      state.treeOpenPageSection = treeOpenSection
    }
  },
  updateSections (state: ApplicationState, sections: {[key:number]: Section}) {
    state.sections = sections
  },
  updateSubSections (state: ApplicationState, subSections: string[]) {
    state.subSectionList = subSections
  },
  updateSectionTree (state: ApplicationState, sections: TreeParentInternal[]) {
    state.treeStructure = sections
  },
  updateVariables (state: ApplicationState, variables: {[key:number]: Variable}) {
    state.variables = variables
  },
  updateGridVariables (state: ApplicationState, gridVariables: VariableWithVariants[]) {
    state.gridVariables = gridVariables
  },
  updateAssessments (state: ApplicationState, assessments: { [key:number]: Assessment }) {
    state.assessments = assessments
  },
  updateVariantCounts (state: ApplicationState, variantCounts: Count[]) {
    state.variantCounts = variantCounts
  },
  updateParticipantCount: (state: ApplicationState, participantCount: number | null) => {
    state.participantCount = participantCount
  },
  updateGridSelection (state: ApplicationState, gridSelection: GridSelection) {
    state.gridSelection = gridSelection
  },
  updateSearchTerm (state: ApplicationState, searchTerm: string|null) {
    state.searchTerm = searchTerm
  },
  updateFilteredSections (state: ApplicationState, sections: number[]) {
    state.filteredSections = sections
  },
  updateFilteredSubsections (state: ApplicationState, subsections: number[]) {
    state.filteredSubsections = subsections
  },
  setTreeCount (state: ApplicationState, count: number) {
    const item:TreeChild|undefined = state.treeStructure[state.treeOpenPageSection - 1].list.find((item:TreeChild) => item.id === state.treeSelected)
    if (item) {
      item.count = count
    }
  },
  toggleGridColumn ({ gridSelection, gridVariables }: {gridSelection: GridSelection, gridVariables: Variable[]}, { assessmentId } : {assessmentId: number}) {
    const allSelected = gridVariables.every((variable) => gridSelection.hasOwnProperty(variable.id) && gridSelection[variable.id].includes(assessmentId))
    if (allSelected) {
      gridVariables.forEach((variable) => {
        const selectedAssessments = gridSelection[variable.id]
        const assessmentIndex = selectedAssessments.indexOf(assessmentId)
        if (assessmentIndex >= 0) {
          selectedAssessments.splice(assessmentIndex, 1)
        }
      })
    } else {
      gridVariables.forEach((variable) => {
        if (!gridSelection.hasOwnProperty(variable.id)) {
          Vue.set(gridSelection, variable.id, [assessmentId])
        } else {
          const selectedAssessments = gridSelection[variable.id]
          if (!selectedAssessments.includes(assessmentId)) {
            selectedAssessments.push(assessmentId)
          }
        }
      })
    }
  },
  toggleGridRow ({ gridSelection, treeSelected }: { gridSelection: GridSelection, treeSelected: number }, { variableId, gridAssessments }: {variableId: number, gridAssessments: Assessment[] }) {
    if (gridSelection.hasOwnProperty(variableId) && (gridSelection[variableId].length === gridAssessments.length)) {
      Vue.delete(gridSelection, variableId)
    } else {
      Vue.set(gridSelection, variableId, gridAssessments.map((it) => it.id))
    }
  },
  toggleAll ({ gridSelection, gridVariables, treeSelected, treeStructure }: {gridSelection: GridSelection, gridVariables: VariableWithVariants[], treeSelected: number, treeStructure: object[]}, { gridAssessments }: {gridAssessments: Assessment[] }) {
    // For each variable all assessments are selected
    const allSelected = gridVariables.every((variable) => {
      return gridSelection.hasOwnProperty(variable.id) && (gridSelection[variable.id].length === gridAssessments.length)
    })
    if (allSelected) {
      gridVariables.forEach((variable) => {
        Vue.delete(gridSelection, variable.id)
      })
    } else {
      gridVariables.forEach((variable) => {
        Vue.set(gridSelection, variable.id, gridAssessments.map((it) => it.id))
      })
    }
  },
  toggleGridSelection ({ gridSelection }: { gridSelection: GridSelection },
    { variableId, assessmentId }: { variableId: number, assessmentId: number }) {
    if (!gridSelection.hasOwnProperty(variableId)) {
      Vue.set(gridSelection, variableId, [assessmentId])
    } else {
      const selectedAssessments = gridSelection[variableId]
      const assessmentIndex = selectedAssessments.indexOf(assessmentId)
      if (assessmentIndex >= 0) {
        if (selectedAssessments.length === 1) {
          Vue.delete(gridSelection, variableId)
        } else {
          selectedAssessments.splice(assessmentIndex, 1)
        }
      } else {
        selectedAssessments.push(assessmentId)
      }
    }
  }
}
