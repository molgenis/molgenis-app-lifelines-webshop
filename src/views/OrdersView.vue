<template>
  <div id="orders-view" class="container pt-1">
    <toast-component
      class="toast-component mt-2"
      v-if="toast"
      :type="toast.type"
      :message="toast.message"
      @toastCloseBtnClicked="clearToast">
    </toast-component>
    <h1 id="orders-title">{{'lifelines-webshop-orders-title' | i18n}}</h1>
      <spinner-animation v-if="orders === null"/>
      <table v-else class="table" aria-describedby="orders-title">
        <thead>
          <tr>
            <th scope="col"></th>
            <th scope="col"></th>
            <th scope="col">Title</th>
            <th scope="col">Submitted</th>
            <th scope="col">Project</th>
            <th scope="col">Order</th>
            <th scope="col">Application form</th>
            <th scope="col">State</th>
          </tr>
        </thead>
        <tbody >
          <tr v-for="order in orders" :key="order.id">
            <td><router-link v-if="order.state === 'Draft'" class="btn btn-primary btn-sm" :to="`/shop/${order.id}`"><font-awesome-icon icon="edit" aria-label="edit"/></router-link></td>
            <td><button v-if="order.state === 'Draft'" class="btn btn-danger btn-sm" @click="deleteOrder(order.id)"><font-awesome-icon icon="trash" aria-label="delete"/></button></td>
            <td>{{ order.name }}</td>
            <td>{{ order.submissionDate | moment("LLLL") }}</td>
            <td>{{ order.projectNumber }}</td>
            <td>{{ order.orderId }}</td>
            <td><span v-if="order.applicationForm"><a :href="order.applicationForm.url">{{ order.applicationForm.filename }}<font-awesome-icon icon="download" aria-label="download"/></a></span></td>
            <td><span class="badge badge-pill" :class="badgeClass[order.state]">{{ order.state }}</span></td>
          </tr>
        </tbody>
      </table>
    </div>
</template>

<script>
import Vue from 'vue'
import { library } from '@fortawesome/fontawesome-svg-core'
import { faEdit, faDownload, faTrash } from '@fortawesome/free-solid-svg-icons'
import { FontAwesomeIcon } from '@fortawesome/vue-fontawesome'
import SpinnerAnimation from '../components/animations/SpinnerAnimation.vue'
import ToastComponent from '../components/ToastComponent.vue'
import { mapActions, mapState, mapMutations } from 'vuex'

library.add(faEdit, faDownload, faTrash)

export default Vue.extend({
  components: { FontAwesomeIcon, SpinnerAnimation, ToastComponent },
  mounted () {
    this.loadOrders()
  },
  data () {
    return {
      badgeClass: {
        'Draft': 'badge-info',
        'Submitted': 'badge-primary',
        'Approved': 'badge-success',
        'Rejected': 'badge-danger'
      }
    }
  },
  methods: {
    ...mapActions(['loadOrders', 'deleteOrder']),
    ...mapMutations(['clearToast'])
  },
  computed: {
    ...mapState(['orders', 'toast'])
  }
})
</script>
