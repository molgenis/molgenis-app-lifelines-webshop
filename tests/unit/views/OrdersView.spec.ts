import { shallowMount, createLocalVue } from '@vue/test-utils'
import VueRouter from 'vue-router'
import OrdersView from '@/views/OrdersView.vue'
import moment from 'moment'
import Vuex from 'vuex'
import orders from '../fixtures/orders'
import Spinner from '@/components/animations/SpinnerAnimation.vue'
import mutations from '@/store/mutations'

describe('OrdersView.vue', () => {
  const localVue = createLocalVue()
  localVue.filter('moment', function (value: string, format: string) { return moment(value).utc().format(format) })
  localVue.filter('i18n', (value: string) => `#${value}#`)

  localVue.use(Vuex)
  localVue.use(VueRouter)
  let store: any
  let actions = {
    loadOrders: jest.fn()
  }

  beforeEach(() => {
    let state: any

    state = {
      orders: null
    }
    store = new Vuex.Store({
      state,
      actions,
      mutations
    })
  })

  it('renders orders view', () => {
    const wrapper = shallowMount(OrdersView, { store, localVue })
    expect(wrapper.find('#orders-view').exists()).toBeTruthy()
  })

  it('shows title', () => {
    const wrapper = shallowMount(OrdersView, { store, localVue })
    expect(wrapper.find('#orders-title').text()).toBe('lifelines-webshop-orders-title')
  })

  it('fetches orders when mounted', () => {
    const wrapper = shallowMount(OrdersView, { store, localVue })
    expect(actions.loadOrders).toHaveBeenCalled()
  })

  it('shows a spinner while loading orders', () => {
    const wrapper = shallowMount(OrdersView, { store, localVue })
    expect(wrapper.find(Spinner).isVisible()).toBeTruthy()
    store.commit('setOrders', orders)
    expect(wrapper.find(Spinner).exists()).toBeFalsy()
  })

  it('shows orders table when loaded', () => {
    const wrapper = shallowMount(OrdersView, { store, localVue })
    store.commit('setOrders', orders)
    expect(wrapper.find('table').isVisible()).toBeTruthy()
  })

  it('checks content of order tables', () => {
    const wrapper = shallowMount(OrdersView, { store, localVue })
    store.commit('setOrders', orders)
    expect(wrapper.find('table')).toMatchSnapshot()
  })
})
