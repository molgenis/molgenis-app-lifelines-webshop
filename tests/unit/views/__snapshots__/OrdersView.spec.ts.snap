// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`OrdersView.vue checks content of order tables 1`] = `
<table aria-describedby="orders-title" class="table">
  <thead>
    <tr>
      <th scope="col"></th>
      <th scope="col"></th>
      <th scope="col">Title</th>
      <th scope="col">Submitted</th>
      <th scope="col">Project</th>
      <th scope="col">Order</th>
      <th scope="col">Application form</th>
      <th scope="col">State</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td>
        <!---->
      </td>
      <td>
        <!---->
      </td>
      <td></td>
      <td>Thursday, October 31, 2019 1:48 PM</td>
      <td></td>
      <td></td>
      <td>
        <!---->
      </td>
      <td><span class="badge badge-pill">3</span></td>
    </tr>
    <tr>
      <td>
        <!---->
      </td>
      <td>
        <!---->
      </td>
      <td>My draft order</td>
      <td>Wednesday, October 30, 2019 1:25 PM</td>
      <td></td>
      <td></td>
      <td><span><a href="https://lifelines.test.molgenis.org/files/aaaac3rcetmgfmudsodb3laaay">Motivation.pdf<font-awesome-icon-stub icon="download" aria-label="download"></font-awesome-icon-stub></a></span></td>
      <td><span class="badge badge-pill">0</span></td>
    </tr>
  </tbody>
</table>
`;
